### PR TITLE
Update on DDStore

### PR DESCRIPTION
**Pull request by jychoi-hpc**
This is the changes we made for using DDStore and the sc23 paper. Mainly, it contains:
* DDStore integration
* Detailed timing during training with GPTL timer
* Adding PNA deg calculation in pre-process and saving in the file
* A few changes

### CODE REVIEW

File: examples/csce/csce_gap.json [hunk 1]
```diff
@@ -47,7 +47,7 @@
       "denormalize_output": false
     },
     "Training": {
-      "num_epoch": 10,
+      "num_epoch": 3,
```
Comment (allaffa) [inline examples/csce/csce_gap.json:50]: Doe we really need to make this change?

File: examples/ising_model/ising_model.json [hunk 2]
```diff
@@ -47,13 +47,13 @@
...
-            "denormalize_output": true
+            "denormalize_output": false
         },
         "Training": {
-            "num_epoch": 2,
+            "num_epoch": 3,
             "perc_train": 0.7,
             "loss_function_type": "mse",
-            "batch_size": 64,
+            "batch_size": 128,
```
Comment (allaffa) [inline examples/ising_model/ising_model.json:56]: Do we really need to keep track of these changes?

File: examples/ising_model/ising_model.json [hunk 3]
```diff
@@ -47,13 +47,13 @@
             "output_names": ["total_energy","spin"],
             "output_index": [0, 1],
             "type": ["graph","node"],
-            "denormalize_output": true
+            "denormalize_output": false
```
Comment (allaffa) [inline examples/ising_model/ising_model.json:50]: Do we really need to keep track of these changes?

File: examples/ising_model/ising_model.json [hunk 4]
```diff
@@ -47,13 +47,13 @@
...
         },
         "Training": {
-            "num_epoch": 2,
+            "num_epoch": 3,
```
Comment (allaffa) [inline examples/ising_model/ising_model.json:53]: Do we really need to keep track of these changes?

File: examples/ising_model/train_ising.py [hunk 5]
```diff
@@ -61,19 +65,32 @@ def write_to_file(total_energy, atomic_features, count_config, dir, prefix):
...
 
 def create_dataset_mpi(
-    L, histogram_cutoff, dir, spin_function=lambda x: x, scale_spin=False, comm=None
+    L,
+    histogram_cutoff,
+    dir,
+    spin_function=lambda x: x,
+    scale_spin=False,
+    comm=None,
+    seed=None,
```
Comment (allaffa) [inline examples/ising_model/train_ising.py:74]: Is the seed used for reproducibility?

File: hydragnn/utils/adiosdataset.py [hunk 6]
```diff
@@ -97,25 +122,35 @@ def save(self):
...
                 m0 = np.min([x.shape for x in arr_list], axis=0)
                 m1 = np.max([x.shape for x in arr_list], axis=0)
-                wh = np.where(m0 != m1)[0]
-                assert len(wh) < 2
-                vdim = wh[0] if len(wh) == 1 else 1
+                vdims = list()
+                for i in range(len(m0)):
+                    if m0[i] != m1[i]:
+                        vdims.append(i)
+                ## We can handle only single variable dimension.
```
Comment (allaffa) [inline hydragnn/utils/adiosdataset.py:135]: What do you mean by single variable dimension?

File: hydragnn/utils/adiosdataset.py [hunk 7]
```diff
@@ -249,11 +313,14 @@ def __init__(
                 self.minmax_node_feature = f.read_attribute(
                     "minmax_node_feature"
                 ).reshape((2, -1))
+            if "pna_deg" in self.attrs:
```
Comment (allaffa) [inline hydragnn/utils/adiosdataset.py:316]: Again, similarly to a previous comment I made both `pna_deg `, can we create a utility function that takes care of this. This attribute is specific to one very specific conversation layer.

File: hydragnn/utils/config_utils.py [hunk 8]
```diff
@@ -39,7 +40,15 @@ def update_config(config, train_loader, val_loader, test_loader):
 
     max_neigh = config["NeuralNetwork"]["Architecture"]["max_neighbours"]
     if config["NeuralNetwork"]["Architecture"]["model_type"] == "PNA":
-        deg = calculate_PNA_degree(train_loader, max_neigh)
+        if "pna_deg" in config:
```
Comment (allaffa) [inline hydragnn/utils/config_utils.py:43]: why do we need this?

**Review by allaffa — COMMENTED** [review]
Most of the comments are actually question to better understand some software design choice.

File: examples/csce/csce_gap.json (hunk 1 above)
Comment (jychoi-hpc) [reply examples/csce/csce_gap.json:50]: Good point. I will go back to the original setting. It was for the SC reproducibility. Now I will leave the branch (`dev-sc23-ddstore`) for the SC reproducibility.

File: examples/ising_model/ising_model.json [hunk 9]
```diff
@@ -23,7 +23,7 @@
         "Architecture": {
             "model_type": "PNA",
             "radius": 7,
-            "max_neighbours": 100000,
+            "max_neighbours": 5,
```
[the remark this answers is not in the document]
Comment (jychoi-hpc) [reply examples/ising_model/ising_model.json:26]: Same here. I will roll back.

File: examples/ising_model/ising_model.json (hunk 3 above)
Comment (jychoi-hpc) [reply examples/ising_model/ising_model.json:50]: Same here. I will roll back.

File: examples/ising_model/ising_model.json (hunk 4 above)
Comment (jychoi-hpc) [reply examples/ising_model/ising_model.json:53]: Same here. I will roll back.

File: hydragnn/utils/adiosdataset.py [hunk 10]
```diff
@@ -17,9 +17,21 @@
...
+    import pyddstore as dds
+except ImportError:
+    pass
+
+import hydragnn.utils.tracer as tr
+
 from hydragnn.utils.abstractbasedataset import AbstractBaseDataset
 
 
+def nsplit(a, n):
```
[the remark this answers is not in the document]
Comment (jychoi-hpc) [reply hydragnn/utils/adiosdataset.py:30]: Will fix.

File: examples/ising_model/train_ising.py (hunk 5 above)
Comment (jychoi-hpc) [reply examples/ising_model/train_ising.py:74]: Yes, it is for reproducibility. I found it is needed when creating on different machines.

File: hydragnn/utils/adiosdataset.py (hunk 6 above)
Comment (jychoi-hpc) [reply hydragnn/utils/adiosdataset.py:131]: The number of variable dimensions. The comment is a bit misleading. I will update. Anyhow, in the code, I am trying to check which dimension is a varying dimension. I can only handle one varying dimension.

File: examples/csce/train_gap.py [hunk 11]
```diff
@@ -304,31 +360,43 @@ def __getitem__(self, idx):
...
+    if args.dataset == "ddstore":
+        os.environ["HYDRAGNN_AGGR_BACKEND"] = "mpi"
+        os.environ["HYDRAGNN_USE_ddstore"] = "1"
+
     (train_loader, val_loader, test_loader,) = hydragnn.preprocess.create_dataloaders(
         trainset, valset, testset, config["NeuralNetwork"]["Training"]["batch_size"]
     )
+    comm.Barrier()
 
+    if hasattr(trainset, "pna_deg"):
```
[the remark this answers is not in the document]
Comment (allaffa) [reply examples/csce/train_gap.py:391]: @jychoi-hpc would you like to discuss this in a meeting?

File: examples/csce/train_gap.py [hunk 12]
```diff
@@ -258,42 +287,69 @@ def __getitem__(self, idx):
...
+        adwriter.add("trainset", trainset)
+        adwriter.add("valset", valset)
+        adwriter.add("testset", testset)
+        adwriter.add_global("pna_deg", deg)
+        adwriter.save()
 
         sys.exit(0)
 
+    tr.initialize()
+    tr.disable()
```
Comment (pzhanggit) [inline examples/csce/train_gap.py:331]: What is the `disable()` for?

File: hydragnn/utils/model.py [hunk 13]
```diff
@@ -89,26 +89,44 @@ def load_existing_model(model, model_name, path="./logs/", optimizer=None):
...
+        return deg
 
 
 def calculate_PNA_degree_dist(loader, max_neighbours):
     assert dist.is_initialized()
     deg = torch.zeros(max_neighbours + 1, dtype=torch.long)
     for data in iterate_tqdm(loader, 2, desc="Calculate PNA degree"):
         d = degree(data.edge_index[1], num_nodes=data.num_nodes, dtype=torch.long)
-        deg += torch.bincount(d, minlength=deg.numel())
+        deg += torch.bincount(d, minlength=deg.numel())[: max_neighbours + 1]
```
Comment (pzhanggit) [inline hydragnn/utils/model.py:111]: Why do we need to explicitly set the length `[: max_neighbours + 1]`?

**Review by pzhanggit — COMMENTED** [review]
Thanks for the PR. I finished first round of review with comments. Will need another round.

File: examples/ising_model/train_ising.py [hunk 14]
```diff
@@ -272,22 +311,58 @@ def info(*args, logtype="info", sep=" "):
...
+            os.path.dirname(__file__), "dataset", "%s.pickle" % modelname
+        )
         trainset = SimplePickleDataset(basedir, "trainset")
         valset = SimplePickleDataset(basedir, "valset")
         testset = SimplePickleDataset(basedir, "testset")
+        minmax_node_feature = trainset.minmax_node_feature
+        minmax_graph_feature = trainset.minmax_graph_feature
+        pna_deg = trainset.pna_deg
+
+        # ## WIP: temporary
```
Comment (allaffa) [inline examples/ising_model/train_ising.py:333]: Are these lines commented with "temporary" still needed for any reason?

File: hydragnn/utils/model.py (hunk 13 above)
Comment (allaffa) [reply hydragnn/utils/model.py:111]: @pzhanggit I think it has to do with extreme cases where the graph has high connectivity, and one node is connected to to so many other nodes within the neighborhood of shell = 1, that the actual number of one distance neighbors exceeds `max_neighbours`. For these extreme cases, you may have to brute-force crop `deg`

**Review by allaffa — COMMENTED** [review]
Second round of review done on my end.

File: examples/csce/train_gap.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (jychoi-hpc) [reply examples/csce/train_gap.py:391]: Thank you for explanation in the last meeting. I am updating now.

File: hydragnn/utils/config_utils.py (hunk 8 above)
Comment (jychoi-hpc) [reply hydragnn/utils/config_utils.py:43]: I am changing this line too as I am moving this "hasattr(trainset, "pna_deg")"  into this `update_config`.

File: examples/csce/train_gap.py [hunk 15]
```diff
@@ -258,42 +287,69 @@ def __getitem__(self, idx):
...
+            use_subdir=True,
+        )
+        SimplePickleWriter(
+            testset,
+            basedir,
+            "testset",
+            use_subdir=True,
+        )
+
+        adwriter = AdiosWriter("examples/csce/dataset/csce_gap.bp", comm)
```
[the remark this answers is not in the document]
Comment (jychoi-hpc) [reply examples/csce/train_gap.py:321]: I will update.

File: examples/ising_model/train_ising.py [hunk 16]
```diff
@@ -28,6 +28,7 @@
 except ImportError:
     pass
 
+import torch_geometric.data
```
[the remark this answers is not in the document]
Comment (jychoi-hpc) [reply examples/ising_model/train_ising.py:31]: You are right. I will remove.

File: hydragnn/utils/adiosdataset.py [hunk 17]
```diff
@@ -85,7 +95,18 @@ def save(self):
...
+                ## If there is no data to save, simply do empty operations
+                ns = self.comm.allgather(len(self.dataset[label]))
+
+                keys_list = self.comm.allgather([])
+                for keys in keys_list:
+                    if len(keys) > 0:
+                        break
+
+                for k in keys:
+                    shape_list = self.comm.allgather([])
```
[the remark this answers is not in the document]
Comment (jychoi-hpc) [reply hydragnn/utils/adiosdataset.py:108]: I found distributed adios writing got an error when some process has no local data to write. Line 99-110 are the routines for doing empty gathering operation to handle the problem. Simple it participates allgather with others but empty contents. 
In short, 
* line 100 is to match with line 111
* line 102 is to match with line 121
* line 108 is to match with line 139

File: examples/csce/train_gap.py (hunk 12 above)
Comment (jychoi-hpc) [reply examples/csce/train_gap.py:331]: Instead of tracing everything right after initialization, I am trying to start tracing only for training loop. I.e., I am disabling tracing here but enabling in the main loop (line 122 in `train_validate_test.py`)

File: examples/ising_model/train_ising.py (hunk 14 above)
Comment (jychoi-hpc) [reply examples/ising_model/train_ising.py:333]: Thanks for checking. It is no use anymore. I will remove.

File: hydragnn/utils/model.py (hunk 13 above)
Comment (pzhanggit) [reply hydragnn/utils/model.py:111]: I made a PR on this https://github.com/ORNL/HydraGNN/pull/176

File: hydragnn/utils/adiosdataset.py [hunk 18]
```diff
@@ -85,7 +95,18 @@ def save(self):
         self.writer = self.io.Open(self.filename, ad2.Mode.Write, self.comm)
         total_ns = 0
         for label in self.dataset:
-            if len(self.dataset[label]) < 1:
+            if len(self.dataset[label]) == 0:
+                ## If there is no data to save, simply do empty operations as follows.
+                ns = self.comm.allgather(len(self.dataset[label]))
+
+                keys_list = self.comm.allgather([])
```
Comment (pzhanggit) [inline hydragnn/utils/adiosdataset.py:102]: I am still not quite following lines 102-108. Are they equivalent to `keys_list=[]*self.size; keys=[]` and `shape_list` will not be assigned values since `keys` is an empty list?

File: hydragnn/utils/adiosdataset.py [hunk 19]
```diff
@@ -97,25 +118,35 @@ def save(self):
                 data = self.dataset[label][0]
                 self.io.DefineAttribute("%s/keys" % label, data.keys)
                 keys = sorted(data.keys)
+                self.comm.allgather(keys)
```
Comment (pzhanggit) [inline hydragnn/utils/adiosdataset.py:121]: Should it be something like `keyslist = self.comm.allgather(keys)`?

File: hydragnn/utils/adiosdataset.py (hunk 18 above)
Comment (allaffa) [reply hydragnn/utils/adiosdataset.py:102]: I think that `keys_list=[]*number_of_processes` is indeed the same of what happens in line 102. 

[https://mpitutorial.com/tutorials/mpi-scatter-gather-and-allgather/](url)

File: hydragnn/utils/adiosdataset.py [hunk 20]
```diff
@@ -85,7 +95,18 @@ def save(self):
...
+                ## If there is no data to save, simply do empty operations as follows.
+                ns = self.comm.allgather(len(self.dataset[label]))
+
+                keys_list = self.comm.allgather([])
+                for keys in keys_list:
+                    if len(keys) > 0:
+                        break
+
+                for k in keys:
+                    shape_list = self.comm.allgather([])
```
Comment (allaffa) [inline hydragnn/utils/adiosdataset.py:108]: To me it seems like `shape_list` gets overwritten every time across successive iterations of the for loop? May you comment on this please?

**Review by allaffa — COMMENTED** [review]
I left a question about `adiosdataset.py`

File: hydragnn/utils/adiosdataset.py (hunk 18 above)
Comment (pzhanggit) [reply hydragnn/utils/adiosdataset.py:102]: Thanks for explaining this in meeting. It is clear to me now.

File: hydragnn/utils/adiosdataset.py (hunk 19 above)
Comment (pzhanggit) [reply hydragnn/utils/adiosdataset.py:121]: Thanks for explaining this in meeting. Clear now.

File: hydragnn/utils/config_utils.py [hunk 21]
```diff
@@ -39,7 +40,15 @@ def update_config(config, train_loader, val_loader, test_loader):
...
     max_neigh = config["NeuralNetwork"]["Architecture"]["max_neighbours"]
     if config["NeuralNetwork"]["Architecture"]["model_type"] == "PNA":
-        deg = calculate_PNA_degree(train_loader, max_neigh)
+        if hasattr(train_loader.dataset, "pna_deg"):
+            deg_bincount = torch.tensor(train_loader.dataset.pna_deg)
+            deg = torch.zeros(max_neigh + 1, dtype=torch.long)
+            if len(deg) < len(deg_bincount):
+                deg[:] = deg_bincount[: len(deg)]
+            else:
+                deg[: len(deg_bincount)] = deg_bincount[:]
```
[the remark this answers is not in the document]
Comment (jychoi-hpc) [reply hydragnn/utils/config_utils.py:61]: I just merged the main branch and now this part is gone.

File: hydragnn/train/train_validate_test.py [hunk 22]
```diff
@@ -100,6 +102,8 @@ def train_validate_test(
     if "Profile" in config:
         profiler.setup(config["Profile"])
 
+    _, rank = get_comm_size_and_rank()
```
[the remark this answers is not in the document]
Comment (jychoi-hpc) [reply hydragnn/train/train_validate_test.py:105]: I will remove.

File: hydragnn/preprocess/utils.py [hunk 23]
```diff
@@ -169,3 +169,21 @@ def __call__(self, data):
...
+    from mpi4py import MPI
+    from hydragnn.utils.print_utils import iterate_tqdm
+
+    max_deg = 0
+    for data in iterate_tqdm(dataset, 2, desc="Degree max"):
+        d = degree(data.edge_index[1], num_nodes=data.num_nodes, dtype=torch.long)
+        max_deg = max(max_deg, max(d))
+    max_deg = MPI.COMM_WORLD.allreduce(max_deg, op=MPI.MAX)
+
+    deg = torch.zeros(max_deg + 1, dtype=torch.long)
```
[the remark this answers is not in the document]
Comment (jychoi-hpc) [reply hydragnn/preprocess/utils.py:184]: Yes. I will remove `+1`

File: hydragnn/utils/config_utils.py [hunk 24]
```diff
@@ -52,7 +52,15 @@ def update_config(config, train_loader, val_loader, test_loader):
...
```
Comment (pzhanggit) [inline hydragnn/utils/config_utils.py:61]:
```suggestion
            deg = torch.tensor(train_loader.dataset.pna_deg)
```

Comment (jychoi-hpc) [reply hydragnn/utils/config_utils.py]: I made some changes. How do you think?

Comment (pzhanggit) [reply hydragnn/utils/config_utils.py]: It looks good to me, thanks!

File: hydragnn/utils/config_utils.py [hunk 25]
```diff
@@ -40,19 +40,12 @@ def update_config(config, train_loader, val_loader, test_loader):
...
-    config["NeuralNetwork"]["Architecture"]["max_neighbours"] = max_degree
-
-    max_neigh = config["NeuralNetwork"]["Architecture"]["max_neighbours"]
     if config["NeuralNetwork"]["Architecture"]["model_type"] == "PNA":
-        deg = calculate_PNA_degree(train_loader, max_neigh)
+        if hasattr(train_loader.dataset, "pna_deg"):
+            ## Use max neibours used in the dataset.
+            deg = torch.tensor(train_loader.dataset.pna_deg)
+        else:
+            deg = gather_deg(train_loader.dataset)
```
[the remark this answers is not in the document]
Comment (jychoi-hpc) [reply hydragnn/utils/config_utils.py:48]: Yes, I added.

[the remark this answers is not in the document]
Comment (pzhanggit) [reply hydragnn/utils/config_utils.py:48]: Thanks!

**Review by pzhanggit — APPROVED** [review]

**Review by allaffa — APPROVED** [review]
